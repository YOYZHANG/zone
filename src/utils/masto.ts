import { login } from 'masto'


function createMasto() {
  return login({
    url: `https://mastodon.social`,
    accessToken: '',
  })
}

export const masto = await createMasto()
