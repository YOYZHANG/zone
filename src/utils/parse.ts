import {parse, TEXT_NODE} from 'ultrahtml'
import type { Node } from 'ultrahtml'

export function htmlToText(html: string | undefined | null) {
  if (!html) return ''
  const ast = parse(html)
  return (ast.children as Node[]).map((n) => treeToText(n)).join('').trim()
}

export function decodeHtml(text: string) {
  const decoder = document.createElement('textarea')
  decoder.innerHTML = text
  return decoder.value
}

export function treeToText(input: Node): string {
  let pre = ''
  let body = ''

  if (input.type === TEXT_NODE)
    return decodeHtml(input.value)

  if (input.name === 'br')
    return '\n'

  if (['p', 'pre'].includes(input.name))
    pre = '\n'

  if ('children' in input)
    body = (input.children as Node[]).map(n => treeToText(n)).join('')

  return pre + body
}
