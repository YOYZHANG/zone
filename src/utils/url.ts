export function getDataUrlFromArr(arr: Uint8ClampedArray, w: number, h: number) {
  if (!w || !h) {
    w = h = Math.sqrt(arr.length/4)
  }

  const canvas = document.createElement('canvas')
  const ctx = canvas.getContext('2d')!

  canvas.width = w
  canvas.height = h

  const imgData = ctx.createImageData(w, h)
  imgData.data.set(arr)
  ctx.putImageData(imgData, 0, 0)

  return canvas.toDataURL()
}
