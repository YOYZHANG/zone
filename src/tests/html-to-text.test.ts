/**
 * @vitest-environment jsdom
 */
import { describe, expect, it } from 'vitest'
import { htmlToText } from '../utils/parse'

describe('html-to-text', () => {
  it('p', () => {
    expect(htmlToText('<p>text inline</p>'))
      .toMatchSnapshot()
  })

  it('p & pre', () => {
    expect(htmlToText('<p>text </p><pre>code</pre>'))
      .toMatchSnapshot()
  })

  it('bold & italic', () => {
    expect(htmlToText('<p>text <b>bold</b> <em>italic</em></p>'))
      .toMatchSnapshot()
  })
})
