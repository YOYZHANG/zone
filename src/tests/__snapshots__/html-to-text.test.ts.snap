// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html-to-text > bold & italic 1`] = `"text **bold** *italic*"`;

exports[`html-to-text > p & pre 1`] = `
"text 
code"
`;

exports[`html-to-text > p 1`] = `"text inline"`;
